pithos/plugins/_mpris.py

```diff
@@ -31,55 +35,86 @@ def __init__(self, window):
-                               dbus.Dictionary({"Metadata": self.__metadata},
-                                               "sv",
-                                               variant_level=1),
-                               [])
-
-    def song_changed(self, artists=None, album=None, title=None, artUrl='',
-                     rating=None):
-        """song_changed - sets the info for the current song.
+            play_state = 'Paused'
+
+        if self._playback_status != play_state:#stops unneeded updates
+            self._playback_status = play_state
+            d = dbus.Dictionary({"PlaybackStatus":self._playback_status},
+                                "sv",variant_level=1)
+            self.PropertiesChanged("org.mpris.MediaPlayer2.Player",d,[])
+
+    def _volumechange_handler(self, window, volume):
+        """Updates the volume in the Sound Menu"""
+
+        if self._volume != volume:#stops unneeded updates
+            self._volume = volume
+            d = dbus.Dictionary({"Volume": self._volume}, "sv",variant_level=1)
+            self.PropertiesChanged("org.mpris.MediaPlayer2.Player",d,[])
+       
+    def _metadatachange_handler(self, window, song):
+        """Updates the song info in the Sound Menu"""
+
+        if song is self.window.current_song:#we only care about the current song
+            self.PropertiesChanged(self.MEDIA_PLAYER2_PLAYER_IFACE,
+            {'Metadata': dbus.Dictionary(self._update_metadata(song),
+            signature='sv'), }, [])
+
+    def _unique_trackid_generator(self, window, song):
+        """The trackid must be a unique and valid object path for each song, although the path does not
+        have to actually point to anything. We use the song.index as our starting point for our trackid
+        so that later when clients refer to songs by their trackid we can extract the song.index for use
+        in Pithos for things like skipping to, or rating a specific song in a tracklist.
+
+        https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-marshaling-object-path
+        https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html#Property:Metadata
+        https://specifications.freedesktop.org/mpris-spec/latest/Track_List_Interface.html#Mapping:Metadata_Map
+        """
+        random_string = ''.join(random.SystemRandom().choice(string.ascii_letters + string.digits) for x in range(56))
+        self._unique_trackid = ('/org/mpris/MediaPlayer2/TrackList/%s%s' %(song.index, random_string))
+        self._valid_trackids.append(self._unique_trackid)
+        logging.info("MPRIS trackid for %s by %s: %s"%(song.title, song.artist, self._unique_trackid))
+
+    def _song_index_from_trackid(self, trackid):
+        """Extracts the song.index from a trackid by slicing the object path
+        from the begining and the random numbers from the end.
+        """
+        return int(trackid[34:-56])
```

Comment:
The string is the MPRIS trackId. It contains the required '/org/mpris/MediaPlayer2/TrackList/' + song_index + a random string of letters and numbers. The trackid must be unique as per the MPRIS docs. It's how MPRIS clients refer to the song. For example ' /org/mpris/MediaPlayer2/TrackList/1D5VDR0XZ6PDlfoD4QppsZesf6AXOjK0hCLzVDPIgHqARoAXcPn3Ekw49'. self._song_index_from_trackid takes the trackid and turns it back to a song index for the purpose of rating the song.